README.md
# flutter_calendar

A new Flutter package project for displaying calendar and the events that are happening currently.

## Getting Started

Add the flutter_calendar dependency to your pubspec.yaml

Add it as a child to any Widget, send size to the widget, by default it occupies the screen width and height

Set the list of events by using static method setListAndUpdateMap present in CalendarEvent class in calender_event.dart file

To add the calendar widget you just need to add below line of code

`CustomCalendar()` as a child

Example:
```
Container(
  child:CustomCalendar(),
);
```

Then you will see the UI as below:


<img src="example/images/no_events.JPG" width="300" height="600" />

If you want to add the events, you need to first set the list of events then build the widget which has CustomCalendar as child(this can be found in example):

```
void setCalendarEvents(){
    List<CalendarEvent> eventsList = List<CalendarEvent>();

    CalendarEvent event = CalendarEvent();
    event.title = "Meeting";
    event.startTime = DateTime(2019,11,01);
    event.endTime = DateTime(2019,11,10);
    event.bgColor = Colors.redAccent;
    eventsList.add(event);

    event = CalendarEvent();
    event.title = "Meeting1";
    event.startTime = DateTime(2019,11,09);
    event.endTime = DateTime(2019,11,09);
    event.bgColor = Colors.redAccent;
    eventsList.add(event);

    event = CalendarEvent();
    event.title = "Meeting2";
    event.startTime = DateTime(2019,11,10);
    event.endTime = DateTime(2019,11,10);
    event.bgColor = Colors.redAccent;
    eventsList.add(event);

    event = CalendarEvent();
    event.title = "Meeting3";
    event.startTime = DateTime(2019,11,01);
    event.endTime = DateTime(2019,11,10);
    event.bgColor = Colors.redAccent;
    eventsList.add(event);

    event = CalendarEvent();
    event.title = "Meeting2";
    event.startTime = DateTime(2019,11,06);
    event.endTime = DateTime(2019,11,15);
    eventsList.add(event);
    CalendarEvent.setListAndUpdateMap(eventsList);
  }
```

UI with events would look like:

<img src="example/images/multipleEventsAndBGColor.png" width="300" height="600" />


CHANGELOG.md
## [1.1.5] - 03-11-2019

* Support for showing custom background color for events
* New UI for showing count if the number of events in the day exceed the count that can be displayed.

## [1.0.5] - 02-07-2019.

* Formatted files 
* decreased font size of day names and color change
* bug fixes in Calendar Event


## [1.0.0] - 02-07-2019.

* A Simple calendar UI that occupies given width and height.
* Shows the list of events on the calendar itself
* Displays events as bars passing through dates for ranged events


example/test/widget_test.dart

import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    
  });
}


test/flutter_calendar_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('test calendar', () {
    
  });
}


lib/month_view.dart
import 'package:auto_size_text/auto_size_text.dart';
import 'package:flutter/material.dart';
import 'calendar_event.dart';
import 'calendar_utils.dart';

final double dateTxtHt = 30;
final double eventItemHt = 20;

class CalendarMonthWidget extends StatefulWidget {
  final DateTime currentMonthDate;
  final Size dayWidgetSize;

  CalendarMonthWidget({
    @required this.currentMonthDate,
    @required this.dayWidgetSize,
  });

  @override
  _CalendarMonthWidgetState createState() => _CalendarMonthWidgetState();
}

class _CalendarMonthWidgetState extends State<CalendarMonthWidget> {
  ///holds the list of events in currentweek
  List<CalendarEvent> eventsInCurrentWeek;

  /// holds the stack positions which are filled w.r.t current day events
  List<int> currentDayEventPositionsInStack = List();

  @override
  Widget build(BuildContext context) {
    int numberOfWeeksInMonth = getNumberOfWeeksInMonth(widget.currentMonthDate);
    return Container(
      child: Column(
        children: <Widget>[
          for (int i = 0; i < numberOfWeeksInMonth; i++) getWeek(i),
        ],
      ),
    );
  }

  Widget getWeek(int weekNumber) {
    int daysBeforeStart = getPaddingBeforeStartDayOfMonth();
    int noOfDaysTillPastWeek = (weekNumber) * 7 - daysBeforeStart;
    setEventsInWeekWithStartDate(noOfDaysTillPastWeek + 1);
    return Container(
      child: createChildren(noOfDaysTillPastWeek + 1),
    );
  }

  ///creates a week view by creating each day's view in a week
  ///
  ///[dayViewWidgets] holds the day widgets which themselves hold the event widgets which happen only on paticular day
  ///
  ///[stackWidgets] holds the event widgets which range accross different dates
  ///
  ///[eventWidgetsInDay] events that happen on single day and also in some cases like the day is first day of week and an event that occurs on more days but end on this day will also be added to this
  
  
  Widget createChildren(int currentDayNumber){
    final List<Widget> dayViewWidgets = [];
    final List<Widget> stackWidgets = [];    
    //creating 7 days
    for(int i = 0;i < 7;i++,currentDayNumber++){      
      final List<Widget> eventWidgetsInDay = [];
      if(currentDayNumber <= 0 || currentDayNumber > getNumberOfDays()){//adding empty views for invalid positions in calendar
        dayViewWidgets.add(getEmptyDay());
      }
      else{
        //get list of events on this date sorted according to their start date and add them to stack or to a dayview          
        final int numberOfEventsToDisplay = (widget.dayWidgetSize.height-dateTxtHt)~/eventItemHt;
        final DateTime currentDay = DateTime(widget.currentMonthDate.year,widget.currentMonthDate.month,currentDayNumber);
        final List<CalendarEvent> sorted = sortedAccordingToTheDuration(currentDay);
        if(numberOfEventsToDisplay != 0){
          for(CalendarEvent event in sorted){      
            final DateTime startDate = DateTime(event.startTime.year,event.startTime.month,event.startTime.day);
            final DateTime endDate = DateTime(event.endTime.year,event.endTime.month,event.endTime.day);
            if(eventWidgetsInDay.length == numberOfEventsToDisplay && eventWidgetsInDay.length >= currentDayEventPositionsInStack.length){
              break;
            }
            if(event.positionInStack >= 0){
              eventWidgetsInDay.add(getEventPlaceHolder());
              continue;
            }
            else if((startDate.difference(currentDay).inDays.abs() > 0 || endDate.difference(currentDay).inDays.abs() > 0) && currentDay.compareTo(DateTime(currentDay.year,currentDay.month,getNumberOfDays())) != 0){
              checkAndAddEventToStack(numberOfEventsToDisplay, event, currentDay, currentDayNumber, i, stackWidgets, eventWidgetsInDay);
            }
            else{
              if(eventWidgetsInDay.length >= numberOfEventsToDisplay)continue;
              else {
                for(int position = 0;position < numberOfEventsToDisplay;position++){
                  if(currentDayEventPositionsInStack.contains(position) || (position < eventWidgetsInDay.length && eventWidgetsInDay.elementAt(position) is EventItem)){
                    //ignoring position if the position is already occupied in stack or if the position already has valid event item widget
                    continue;
                  }
                  eventWidgetsInDay.insert(position,getEventItem(event: event,));
                  break;
                } 
              }
            }
          }
        }
        //added a day with event widgets
        dayViewWidgets.add(getDayWidget(currentDay,eventWidgetsInDay));
        if(sorted.length-numberOfEventsToDisplay > 0){
          const size = 25.0;
          stackWidgets.add(
            Positioned(
              left: i*widget.dayWidgetSize.width,
              top:0,
              width: size,
              child: XmoreWidget(sorted.length-numberOfEventsToDisplay,size: size,),
            )
          );
        }
      }
    }
    return Stack(
      children: <Widget>[
        Row(
          mainAxisSize: MainAxisSize.min,
          children: dayViewWidgets,
        ),
        ...stackWidgets
      ],
    );
  }


/// adds an ranged event to stack by checking the positions that are empty
 void checkAndAddEventToStack(int numberOfEventsToDisplay, CalendarEvent event, DateTime currentDay, int currentDayNumber, int i, List<Widget> stackWidgets, List<Widget> eventWidgetsInDay) {
    for(int position = 0;position < numberOfEventsToDisplay;position++){
      if(currentDayEventPositionsInStack.contains(position)  || (position < eventWidgetsInDay.length && eventWidgetsInDay.elementAt(position) is EventItem)){
        continue;
      }
      currentDayEventPositionsInStack.add(position);
      event.positionInStack = position;
      final int eventDuration = event.endTime.difference(currentDay).inDays+1;
      final int noOfDaysLeftInWeek = (getNumberOfDays() - currentDayNumber)+1 >= (7-i) ? 7 - i:(getNumberOfDays()-currentDayNumber)+1;
      final double width = (eventDuration <= noOfDaysLeftInWeek?eventDuration:noOfDaysLeftInWeek) * widget.dayWidgetSize.width;
      stackWidgets.add(Positioned(
        left: i*widget.dayWidgetSize.width,
        top: position*eventItemHt+dateTxtHt,
        width: width,
        child: IgnorePointer(child: 
        getEventItem(event: event,width: width))
      ));
      eventWidgetsInDay.add(getEventPlaceHolder());                  
      break;//break after the event is added
      // 
    }
  }


  /// returns an empty view - the invalid days at the start and end of the month view with no date in them
  Widget getEmptyDay() {
    return Container(
      decoration: BoxDecoration(
          shape: BoxShape.rectangle,
          border: Border.all(color: Colors.grey[300], width: 0.35),
          color: Colors.white),
      width: widget.dayWidgetSize.width,
      height: widget.dayWidgetSize.height,
      padding: EdgeInsets.only(top: 5),
    );
  }

  /// creates place holders which acts as dummy events -- to avoid overlapping a day events with events that
  /// range between 2 to any no.of days
  Widget getEventPlaceHolder() {
    return SizedBox(
      width: widget.dayWidgetSize.width,
      height: eventItemHt,
    );
  }

  /// return a single event widget that might be added to a day view or else 
/// to stack to display as a continuous UI event through days
  Widget getEventItem({@required CalendarEvent event,double width}){
     return EventItem(
      eventItemHt, 
      width??widget.dayWidgetSize.width,
      event,
    );
  }


/// returns a [day] view(contains date and events on particular date) in a week by adding the [eventWidgets]
  Widget getDayWidget(DateTime day,List<Widget> eventWidgets){
   return InkWell(
      onTap: (){
        // EventsListAlert.showEventsAlert(context,DateTime(day.year,day.month,day.day));
      },
      child: Container(
        decoration: BoxDecoration(shape: BoxShape.rectangle,border: Border.all(color: Colors.grey[300],width: 0.35),color: Colors.white),
        width: widget.dayWidgetSize.width,
        height: widget.dayWidgetSize.height,
        padding: EdgeInsets.only(top: 5),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            getDateWidget(day.day),
            ...eventWidgets
          ],
        )
      )
    );
  }

  /// creates and returns a text view with [date] as the text in it.
  ///
  /// if date is today then it returns a circular widget with [date] as text and background color as #FF1E90FF
  Widget getDateWidget(int date) {
    DateTime now = DateTime.now();
    bool isToday = (now.day == date &&
        now.month == widget.currentMonthDate.month &&
        now.year == widget.currentMonthDate.year);
    return getCircularWidget(
        padding: EdgeInsets.all(0),
        child: Text('$date',
            style: Theme.of(context).textTheme.display1.copyWith(
                color: isToday ? Colors.white : Colors.black, fontSize: 13)),
        fillColor: isToday ? Color(0xFF1E90FF) : Colors.transparent);
  }

  List<CalendarEvent> sortedAccordingToTheDuration(DateTime date) {
    List<CalendarEvent> events = List();
    currentDayEventPositionsInStack =
        List(); //resetting current day positions in stack
    for (CalendarEvent event in eventsInCurrentWeek) {
      DateTime startDate = DateTime(
          event.startTime.year, event.startTime.month, event.startTime.day);
      DateTime endDate =
          DateTime(event.endTime.year, event.endTime.month, event.endTime.day);
      if (date.compareTo(startDate) >= 0 && date.compareTo(endDate) <= 0) {
        if (events.contains(event)) continue;
        events.add(event);
        if (event.positionInStack >= 0) {
          currentDayEventPositionsInStack.add(event.positionInStack);
        }
      }
    }
    events.sort(comparator);
    return events;
  }

  int comparator(CalendarEvent event1, CalendarEvent event2) {
    int compareOutput = event1.startTime.compareTo(event2.startTime);
    if (compareOutput < 0)
      return -1; //makes event1 come before event2 in the result list
    else if (compareOutput > 0)
      return 1; //makes event2 come before event1 in the result list
    else {
      return event1.startTime
          .difference(event1.endTime)
          .inMilliseconds
          .abs()
          .compareTo(
              event2.startTime.difference(event2.endTime).inMilliseconds.abs());
    }
  }

  void setEventsInWeekWithStartDate(int date) {
    int totDays = getNumberOfDays();
    eventsInCurrentWeek = List();
    for (int i = 0; i < 7; i++, date++) {
      if (date <= 0 || date > totDays) continue;
      eventsInCurrentWeek.addAll(getEventsOn(DateTime(
          widget.currentMonthDate.year, widget.currentMonthDate.month, date)));
    }
  }

  List<CalendarEvent> getEventsOn(DateTime date) {
    List<int> eventPositions = CalendarEvent.getList(date.month, date.year);
    List<CalendarEvent> eventsOnDate = List();
    for (int pos in eventPositions) {
      CalendarEvent event = CalendarEvent.eventsList[pos];
      DateTime startDate = DateTime(
          event.startTime.year, event.startTime.month, event.startTime.day);
      DateTime endDate =
          DateTime(event.endTime.year, event.endTime.month, event.endTime.day);
      if (date.compareTo(startDate) >= 0 && date.compareTo(endDate) <= 0) {
        event.positionInStack = -1;
        eventsOnDate.add(event);
      }
    }
    return eventsOnDate;
  }

  int getPaddingBeforeStartDayOfMonth() {
    DateTime dateTime = DateTime(
        widget.currentMonthDate.year, widget.currentMonthDate.month, 1);
    return dateTime.weekday == 7 ? 0 : dateTime.weekday;
  }

  int getNumberOfDays() {
    return getNumberOfDaysInMonth(widget.currentMonthDate);
  }
}
class EventItem extends StatelessWidget {
  const EventItem(this.height,this.width,this.event,{this.horizontalPadding = 4,this.verticalPadding = 4});
  
  final int verticalPadding;
  final int horizontalPadding;

  final double height;
  final double width;
  final CalendarEvent event;

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.symmetric(horizontal: 2),
      padding: EdgeInsets.symmetric(vertical: verticalPadding/2),
      child: Container(
        width: width.toDouble()-horizontalPadding,
        height: (height-verticalPadding).toDouble(),
        color: event.getEventColor(),
        child:Center(child: getTitleWidget(context)),
      )
    );
  }

  Widget getTitleWidget(BuildContext context){
    return Text
    (
      event.title,
      style: Theme.of(context).textTheme.body1.copyWith(fontSize: 14),
      overflow: TextOverflow.ellipsis,
      textAlign: TextAlign.center,
      maxLines: 1,
    );
  }
}



class XmoreWidget extends StatelessWidget {

  const XmoreWidget(this.xmoreVal,{this.size = 25});

  final int xmoreVal;

  final double size;
  
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size(size, size),
      child: SizedBox(
        width: size,
        height: size,
        child: Stack(
          children: <Widget>[
            Padding(
              padding: EdgeInsets.only(top: xmoreVal>9?0:0.5),
              child: Align(
                alignment: Alignment.topLeft,
                child: AutoSizeText(
                  '+$xmoreVal',
                  style: Theme.of(context).textTheme.display1.copyWith(color: Colors.white,fontSize: 11),
                  maxFontSize: 12,
                  minFontSize: 8,
                  maxLines: 2,
                )
              ),
            )
          ],
        ),
      ),
      painter:TrianglePainter() ,
    );
  }

}


class TrianglePainter extends CustomPainter {

  @override
  void paint(Canvas canvas, Size size) {
    
    final Paint paintBrush = Paint();
    paintBrush.color = Color(0xFFffc422).withAlpha(150);

    //reversed triangle
    final reversePath = Path();
    reversePath.lineTo(0, 0);
    reversePath.lineTo(size.width, 0);
    reversePath.lineTo(0, size.width);
    reversePath.lineTo(0, 0);
    reversePath.close();

    canvas.drawPath(reversePath,paintBrush);

  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
    
}


lib/calendar_event.dart

import 'package:flutter/material.dart';

class CalendarEvent{

  String title;
  String extension,path;
  DateTime startTime,endTime;
  Color bgColor;

  /// used while painting event in calendar
  int positionInStack = -1; //current position

  CalendarEvent();
  static List<CalendarEvent> _events;

  static List<CalendarEvent> get eventsList => _events;

  static Map<String, List<int>> _eventsDict;

  static void setListAndUpdateMap(List<CalendarEvent> events) {
    //on deleting an event
    _events = events;
    _eventsDict = Map();
    for (int i = 0; i < events.length; i++) {
      CalendarEvent event = events.elementAt(i);
      _updateMap(event, i);
    }
  }

  //indexing events for fast processing
  static void _updateMap(CalendarEvent event, int position) {
    int monthsCount = numberOfMonthsBetween(event.startTime, event.endTime);
    int index = 0;
    do {
      DateTime dateTime =
          DateTime(event.startTime.year, event.startTime.month + index, 1);
      String key = _getKeyFrom(dateTime.month, dateTime.year);
      List<int> pointersToEvents = _eventsDict[key];
      if (pointersToEvents == null) {
        pointersToEvents = List<int>();
      }
      pointersToEvents.add(position);
      _eventsDict[key] = pointersToEvents;
      index++;
    } while (index < monthsCount);
  }

  static List<int> getList(int month, int year) {
    if (_eventsDict == null) return List<int>();
    List<int> eventPositions = _eventsDict[_getKeyFrom(month, year)];
    return eventPositions ?? List<int>();
  }

  static String _getKeyFrom(int month, int year) {
    return '$month-$year';
  }

  bool isImageType() {
    return (extension != null &&
        (extension == "jpeg" ||
            extension == "jpg" ||
            extension == "png" ||
            extension == "gif"));
  }

  Color getEventColor() {
    return bgColor??Colors.grey;
  }
}

int numberOfMonthsBetween(DateTime startDate, DateTime endDate,
    {bool inclusiveStartAndEnd = true}) {
  int numberOfYearsInBetween = endDate.year - startDate.year;
  if (numberOfYearsInBetween < 0) {
    numberOfYearsInBetween = numberOfYearsInBetween.abs();
    DateTime temp = startDate;
    startDate = endDate;
    endDate = temp;
  }
  if (numberOfYearsInBetween == 0) {
    int noOfMonths = endDate.month - startDate.month;
    return inclusiveStartAndEnd
        ? noOfMonths + 1
        : noOfMonths > 0 ? noOfMonths - 1 : 0;
  } else if (numberOfYearsInBetween > 0) {
    int noOfMonths = 12 -
        startDate.month +
        endDate.month +
        12 * (numberOfYearsInBetween - 1);
    return inclusiveStartAndEnd ? noOfMonths + 1 : noOfMonths - 1;
  }
  return -1;
}


lib/calendar_utils.dart
import 'package:flutter/material.dart';

int getNumberOfWeeksInMonth(DateTime date) {
  DateTime dateTime = DateTime(date.year, date.month, 1);
  int noOfDays = getNumberOfDaysInMonth(date);
  int noOfDaysInStartWeek = 7 - dateTime.weekday;
  int numberOfWeeks = (noOfDays - noOfDaysInStartWeek) ~/
      7; // ~/ --> truncateing integer divison opertor
  numberOfWeeks += (noOfDays - noOfDaysInStartWeek) % 7 != 0 ? 1 : 0;
  if (noOfDaysInStartWeek > 0)
    numberOfWeeks++; //add a week for the noOfDaysInStartWeek
  // Logger.log("$numberOfWeeks -- number of weeks");
  return numberOfWeeks;
}

int getNumberOfDaysInMonth(DateTime date) {
  DateTime dateTime = DateTime(date.year, date.month, 1);
  DateTime nextMonth = DateTime(date.year, date.month + 1, 1);
  Duration duration = nextMonth.difference(dateTime);
  // Logger.log("${duration.inDays} -- number of days \n ${dateTime.weekday} -- weekday");
  return duration.inDays;
}

Widget getCircularWidget(
    {Widget child,
    Color fillColor: Colors.white,
    double size: 24,
    EdgeInsetsGeometry margin: const EdgeInsets.only(left: 0, right: 0),
    EdgeInsetsGeometry padding: const EdgeInsets.all(4.0)}) {
  return Container(
    margin: margin,
    width: size,
    height: size,
    decoration: BoxDecoration(shape: BoxShape.circle, color: fillColor),
    child: Center(
      child: child != null
          ? Padding(padding: padding, child: child)
          : Padding(
              padding: EdgeInsets.all(0),
            ),
    ),
  );
}


lib/calendar_flutter.dart
library calendar_flutter;

import 'dart:async';
import 'dart:ui';
import 'package:flutter/material.dart';

import 'calendar_utils.dart';
import 'month_view.dart';

class CustomCalendar extends StatefulWidget {
  final List<String> weekDays;
  final Size calendarSize;
  CustomCalendar({this.weekDays, this.calendarSize});
  @override
  _CustomCalendarState createState() => _CustomCalendarState();
}

class _CustomCalendarState extends State<CustomCalendar>
    with SingleTickerProviderStateMixin {
  final int numWeekDays = 7;
  Size size;
  double itemHeight;
  double itemWidth;
  DateTime _currentDate = DateTime.now();
  PageController _controller;
  int prevIndex = 12;
  StreamController<int> _dateStreamController = StreamController();

  @override
  void initState() {
    _controller = PageController(
      initialPage: prevIndex,
      keepPage: false,
      viewportFraction: 1.0,
    );
    super.initState();
  }

  @override
  void dispose() {
    _dateStreamController.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    initDimensions();
    return Container(
        color: Colors.white, child: buildCalendarWithEvents(context));
  }

  Column buildCalendarWithEvents(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        getWeekDaysView(),
        Expanded(child: monthBuilder()),
      ],
    );
  }

/////////////////////////////////////////
  ///--------UI methods----------//////
/////////////////////////////////////////

  /// creates widget with list of days from Sunday to Saturday in a row.
  Widget getWeekDaysView() {
    return Container(
      padding: EdgeInsets.symmetric(vertical: 12.0),
      color: Color(0x551A609F),
      child: Row(
        children: <Widget>[
          for (String day in getWeekDays())
            SizedBox(
                width: itemWidth,
                child: Container(
                    padding: EdgeInsets.symmetric(horizontal: 4),
                    child: Text(
                      day,
                      style: Theme.of(context)
                          .textTheme
                          .body1
                          .copyWith(color: Colors.black),
                      overflow: TextOverflow.ellipsis,
                      maxLines: 1,
                    )))
        ],
      ),
    );
  }

  Widget monthBuilder() {
    return PageView.builder(
        controller: _controller,
        onPageChanged: (index) {
          setCurrentDate(index);
        },
        itemBuilder: (context, index) {
          setCurrentDate(index);
          initDimensions();
          return CalendarMonthWidget(
            dayWidgetSize: Size(itemWidth, itemHeight),
            currentMonthDate: _currentDate,
          );
        });
  }

  void setCurrentDate(int index) {
    int month = _currentDate.month;
    if (index > prevIndex) {
      month += index - prevIndex;
    } else if (index < prevIndex) {
      month -= prevIndex - index;
    }
    prevIndex = index;
    _currentDate = DateTime(_currentDate.year, month, 1);
    _dateStreamController.add(0); //just to notify builder
  }

/////////////////////////////////////////
  ///--------helper methods----------//////
/////////////////////////////////////////

  List<String> getWeekDays() {
    return widget.weekDays ??
        ['Sun', 'Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat'];
  }

  void initDimensions() {
    if (widget.calendarSize == null) {
      size = MediaQuery.of(context).size;
      itemHeight =
          (size.height - kBottomNavigationBarHeight - kToolbarHeight - 80) /
              getNumberOfWeeksInMonth(_currentDate);
      itemWidth = size.width / numWeekDays;
    } else {
      size = widget.calendarSize;
      itemHeight =
          widget.calendarSize.width / getNumberOfWeeksInMonth(_currentDate);
      itemWidth = size.width / numWeekDays;
    }
  }
}


lib/flutter_calendar.dart
library flutter_calendar;

import 'dart:async';
import 'dart:ui';
import 'package:flutter/material.dart';

import 'calendar_utils.dart';
import 'month_view.dart';

class CustomCalendar extends StatefulWidget {
  final List<String> weekDays;
  final Size calendarSize;
  CustomCalendar({this.weekDays, this.calendarSize});
  @override
  _CustomCalendarState createState() => _CustomCalendarState();
}

class _CustomCalendarState extends State<CustomCalendar>
    with SingleTickerProviderStateMixin {
  final int numWeekDays = 7;
  Size size;
  double itemHeight;
  double itemWidth;
  DateTime _currentDate = DateTime.now();
  PageController _controller;
  int prevIndex = 3;
  StreamController<int> _dateStreamController = StreamController();

  @override
  void initState() {
    _controller = PageController(
      initialPage: prevIndex,
      keepPage: false,
      viewportFraction: 1.0,
    );
    super.initState();
  }

  @override
  void dispose() {
    _dateStreamController.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    initDimensions();
    return Container(
        color: Colors.white, child: buildCalendarWithEvents(context));
  }

  Column buildCalendarWithEvents(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        getWeekDaysView(),
        Expanded(child: monthBuilder()),
      ],
    );
  }

/////////////////////////////////////////
  ///--------UI methods----------//////
/////////////////////////////////////////

  /// creates widget with list of days from Sunday to Saturday in a row.
  Widget getWeekDaysView() {
    return Container(
      padding: EdgeInsets.symmetric(vertical: 12.0),
      color: Color(0x551A609F),
      child: Row(
        children: <Widget>[
          for (String day in getWeekDays())
            SizedBox(
                width: itemWidth,
                child: Container(
                    padding: EdgeInsets.symmetric(horizontal: 4),
                    child: Text(
                      day,
                      style: Theme.of(context)
                          .textTheme
                          .body1
                          .copyWith(color: Colors.black),
                      overflow: TextOverflow.ellipsis,
                      maxLines: 1,
                    )))
        ],
      ),
    );
  }

  Widget monthBuilder() {
    return PageView.builder(
        controller: _controller,
        onPageChanged: (index) {
          setCurrentDate(index);
        },
        itemBuilder: (context, index) {
          setCurrentDate(index);
          initDimensions();
          return CalendarMonthWidget(
            dayWidgetSize: Size(itemWidth, itemHeight),
            currentMonthDate: _currentDate,
          );
        });
  }

  void setCurrentDate(int index) {
    int month = _currentDate.month;
    if (index > prevIndex) {
      month += index - prevIndex;
    } else if (index < prevIndex) {
      month -= prevIndex - index;
    }
    prevIndex = index;
    _currentDate = DateTime(_currentDate.year, month, 1);
    _dateStreamController.add(0); //just to notify builder
  }

/////////////////////////////////////////
  ///--------helper methods----------//////
/////////////////////////////////////////

  List<String> getWeekDays() {
    return widget.weekDays ??
        ['Sun', 'Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat'];
  }

  void initDimensions() {
    if (widget.calendarSize == null) {
      size = MediaQuery.of(context).size;
      itemHeight =
          (size.height - kBottomNavigationBarHeight - kToolbarHeight - 80) /
              getNumberOfWeeksInMonth(_currentDate);
      itemWidth = size.width / numWeekDays;
    } else {
      size = widget.calendarSize;
      itemHeight =
          widget.calendarSize.width / getNumberOfWeeksInMonth(_currentDate);
      itemWidth = size.width / numWeekDays;
    }
  }
}


example/lib/main.dart
import 'package:calendar_flutter/calendar_event.dart';
import 'package:calendar_flutter/flutter_calendar.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  

  @override
  Widget build(BuildContext context) {
    //setCalendarEvents();// uncomment this to see the events on the calendar
    return Scaffold(
      appBar: AppBar(
        title: Text('Calendar'),
      ),
      body:CustomCalendar()
    );
  }

  void setCalendarEvents(){
    List<CalendarEvent> eventsList = List<CalendarEvent>();

    CalendarEvent event = CalendarEvent();
    event.title = "Meeting";
    event.startTime = DateTime(2019,07,01);
    event.endTime = DateTime(2019,07,10);
    eventsList.add(event);

    event = CalendarEvent();
    event.title = "Meeting2";
    event.startTime = DateTime(2019,07,06);
    event.endTime = DateTime(2019,07,15);
    eventsList.add(event);
    CalendarEvent.setListAndUpdateMap(eventsList);
  }
}
